### PR TITLE
Relayer: Refactor the relayer main loop

### PR DESCRIPTION
**Pull request by sczembor**
The `btc` processing logic moved to a separate package. Now the relayer uses two processors: `nativeProcessor` and `btcProcessor` + updated better tests + couple of minor fixes/cleanups

Closes: #54 

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
    
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary

### CODE REVIEW

File: ika2btc/processor.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+			}
+			log.Info().Msgf("Transaction confirmed: %s", tx.BtcTxID)
+		}
+	}
+	return nil
+}
+
+// Shutdown shuts down the Bitcoin RPC client.
+func (p *Processor) Shutdown() {
+	p.BtcClient.Shutdown()
```
Comment (vuvoth) [inline ika2btc/processor.go:131]: Should we close db connection also?

Comment (vuvoth) [reply ika2btc/processor.go:131]: btcClient here is the full node right?

Comment (sczembor) [reply ika2btc/processor.go:131]: The relayer is responsible for closing the database connection, this is just the `processor`. So relayer will call `processor.Shutdown()` and then `db.close()` when stoping itself

File: ika2btc/processor.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+func NewProcessor(
+	btcClientConfig rpcclient.ConnConfig,
+	confirmationThreshold uint8,
+	db *dal.DB,
+) (*Processor, error) {
+
+	if db == nil {
+		err := fmt.Errorf("database cannot be nil")
+		log.Err(err).Msg("")
+		return nil, err
```
Comment (robert-zaremba) [inline ika2btc/processor.go:34]: Error logging should be in the main package ideally. Debug / info should be "locally". Also, let's move the error construction to a  `const`.
```suggestion
		return nil, ErrNoDB
```

Comment (robert-zaremba) [reply ika2btc/processor.go:34]: And error in `errors.go`:
```go
const ErrNoDB = errors.New("database can not be nil")
```

File: ika2btc/processor.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	if btcClientConfig.Host == "" || btcClientConfig.User == "" || btcClientConfig.Pass == "" {
+		err := fmt.Errorf("missing bitcoin node configuration")
+		log.Err(err).Msg("")
```
Comment (robert-zaremba) [inline ika2btc/processor.go:39]: same as above

File: ika2btc/processor.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+	return &Processor{
+		BtcClient:               client,
+		db:                      db,
+		txConfirmationThreshold: confirmationThreshold,
+	}, nil
+}
+
+// ProcessSignedTxs processes signed transactions from the database.
+func (p *Processor) ProcessSignedTxs(mu *sync.Mutex) error {
```
Comment (robert-zaremba) [inline ika2btc/processor.go:56]:
```suggestion
// Run starts a loop to query and process signed transactions from the database.
func (p *Processor) Run(mu *sync.Mutex) error {
```

File: ika2btc/processor.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
+package ika2btc
```
Comment (robert-zaremba) [inline ika2btc/processor.go:1]: If this package doesn't interact with `ika` then let's rename it - it shouldn't have `ika` in the prefix.

File: ika2btc/processor.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+		BtcClient:               client,
+		db:                      db,
+		txConfirmationThreshold: confirmationThreshold,
+	}, nil
+}
+
+// ProcessSignedTxs processes signed transactions from the database.
+func (p *Processor) ProcessSignedTxs(mu *sync.Mutex) error {
+	mu.Lock()
+	defer mu.Unlock()
```
Comment (robert-zaremba) [inline ika2btc/processor.go:58]: let's remove mutex now, we will do it in a different way.

File: ika2btc/processor_test.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+var btcClientConfig = rpcclient.ConnConfig{
+	Host:         "test_rpc",
+	User:         "test_user",
+	Pass:         "test_pass",
+	HTTPPostMode: true,
+	DisableTLS:   false,
+}
+
+func TestProcessor_ProcessSignedTxs(t *testing.T) {
```
Comment (robert-zaremba) [inline ika2btc/processor_test.go:22]:
```suggestion
func TestProcessSignedTxs(t *testing.T) {
```

File: ika2btc/processor_test.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+	var mu sync.Mutex
+	err = processor.ProcessSignedTxs(&mu)
+	assert.NilError(t, err)
+
+	updatedTx, err := db.GetBitcoinTx(2, daltest.GetHashBytes(t, "0"))
+	assert.NilError(t, err)
+	assert.Equal(t, updatedTx.Status, dal.Broadcasted)
+}
+
+func TestProcessor_CheckConfirmations(t *testing.T) {
```
Comment (robert-zaremba) [inline ika2btc/processor_test.go:39]:
```suggestion
func TestCheckConfirmations(t *testing.T) {
```

File: ika2btc/processor_test.go [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+	assert.Assert(t, processor == nil)
+}
+
+func TestNewProcessor_MissingBtcConfig(t *testing.T) {
+	db := daltest.InitTestDB(t)
+	btcClientConfig.Host = ""
+	processor, err := NewProcessor(btcClientConfig, 6, db)
+	assert.ErrorContains(t, err, "missing bitcoin node configuration")
+	assert.Assert(t, processor == nil)
+}
```
Comment (robert-zaremba) [inline ika2btc/processor_test.go:68]:
```suggestion
func TestNewProcessor(t *testing.T) {
	// missing db
	processor, err := NewProcessor(btcClientConfig, 6, nil)
	assert.ErrorIs(t, err, ErrNoDB)
	assert.Assert(t, processor)
	
	// missing BTC config
	db := daltest.InitTestDB(t)
	btcClientConfig.Host = ""
	processor, err = NewProcessor(btcClientConfig, 6, db)
	assert.ErrorIs(t, err, ErrNoBtcConfig)
	assert.Assert(t, processor)
}
```

File: ika2btc/processor.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+		err = p.db.InsertBtcTx(dal.BitcoinTx{
+			SrID:      tx.ID,
+			Status:    dal.Broadcasted,
+			BtcTxID:   txHash.CloneBytes(),
+			Timestamp: time.Now().Unix(),
+			Note:      "",
+		})
+		if err != nil {
+			return fmt.Errorf("DB: can't update tx status: %w", err)
```
Comment (robert-zaremba) [inline ika2btc/processor.go:88]:
```suggestion
			return fmt.Errorf("DB: can't update tx status: {tx: %d, err: %w}", tx.ID, err)
```

Comment (robert-zaremba) [reply ika2btc/processor.go:88]: check if `tx.ID` is a number, otherwise change `%d` to `%s`

File: ika2btc/processor_test.go [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+	DisableTLS:   false,
+}
+
+func TestProcessor_ProcessSignedTxs(t *testing.T) {
+	db := daltest.InitTestDB(t)
+	daltest.PopulateSignRequests(t, db)
+	daltest.PopulateBitcoinTxs(t, db)
+	processor, err := NewProcessor(btcClientConfig, 6, db)
+	assert.NilError(t, err)
+	processor.BtcClient = &bitcoin.MockClient{}
```
Comment (robert-zaremba) [inline ika2btc/processor_test.go:28]:
```suggestion
	db, processor := initProcessor(t)
```

Comment (robert-zaremba) [reply ika2btc/processor_test.go:28]: move setup code to a helper function

File: ika2btc/processor_test.go [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+	assert.Equal(t, updatedTx.Status, dal.Broadcasted)
+}
+
+func TestProcessor_CheckConfirmations(t *testing.T) {
+	db := daltest.InitTestDB(t)
+	daltest.PopulateSignRequests(t, db)
+	daltest.PopulateBitcoinTxs(t, db)
+	processor, err := NewProcessor(btcClientConfig, 6, db)
+	assert.NilError(t, err)
+	processor.BtcClient = &bitcoin.MockClient{}
```
Comment (robert-zaremba) [inline ika2btc/processor_test.go:45]: as above

File: nbtc/relayer_test.go [hunk 13]
```diff
@@ -55,49 +64,35 @@ func Test_Start(t *testing.T) {
...
+	mockIkaClient := ika.NewMockClient()
+	btcProcessor, _ := ika2btc.NewProcessor(btcClientConfig, 6, db)
+	btcProcessor.BtcClient = &bitcoin.MockClient{}
+	nativeProcessor := native2ika.NewProcessor(mockIkaClient, db)
 
-	updatedTx, err := db.GetBitcoinTx(2, daltest.GetHashBytes(t, "0"))
-	assert.NilError(t, err)
-	assert.Equal(t, updatedTx.Status, dal.Broadcasted)
+	relayer, err := NewRelayer(relayerConfig, nil, nativeProcessor, btcProcessor)
+	assert.ErrorContains(t, err, "database cannot be nil")
```
Comment (robert-zaremba) [inline nbtc/relayer_test.go:75]:
```suggestion
	assert.ErrorIs(t, err, dal.ErrNoDB)
```

File: nbtc/relayer_test.go [hunk 14]
```diff
@@ -55,49 +64,35 @@ func Test_Start(t *testing.T) {
...
+	assert.Assert(t, relayer == nil)
 }
 
-func Test_checkConfirmations(t *testing.T) {
+func Test_NewRelayer_NativeProcessorError(t *testing.T) {
```
Comment (robert-zaremba) [inline nbtc/relayer_test.go:79]: let's merge this test with the one above

**Review by robert-zaremba — APPROVED** [review]
pre-approving

File: ika2btc/processor.go (hunk 2 above)
Comment (sczembor) [reply ika2btc/processor.go]: Done

File: ika2btc/processor.go (hunk 3 above)
Comment (sczembor) [reply ika2btc/processor.go:37]: Done

File: ika2btc/processor.go (hunk 6 above)
Comment (sczembor) [reply ika2btc/processor.go:56]: Done

File: nbtc/relayer_test.go (hunk 13 above)
Comment (sczembor) [reply nbtc/relayer_test.go:75]: Done

File: nbtc/relayer_test.go (hunk 14 above)
Comment (sczembor) [reply nbtc/relayer_test.go:79]: Done

File: ika2btc/processor_test.go (hunk 11 above)
Comment (sczembor) [reply ika2btc/processor_test.go]: Done

**Pull request merged**